### PR TITLE
Add optional sortsupport methods to the PostgreSQL GiST indices

### PR DESCRIPTION
**Pull request by rvianello**
#### What does this implement/fix? Explain your changes.

I finally found some time to resume working on (and mostly completed) my RDKit UGM sprints project. I am still running tests and I'm afraid support for bfp data could/should be eventually disabled by default. On the other hand, based on some (very) preliminary data, there might be some useful performance improvement for substructure searches.

PostgreSQL 14 introduced support for pre-sorting the data when building GiST indices. This means that if a suitable comparator function is defined, instead of using the `penalty` and `picksplit` functions to insert the a new entry starting from the root node, the destination leaf node can be determined upfront, and the index tree is then updated bottom-up following the direct insertion of the new entry into the corresponding leaf node.

In general, pre-sorting is expected to reduce the create and reindex execution time and allow the construction of more compact indices. In the case of the RDKit GiST indices I think the `penalty` and `picksplit` implementation were always a problematic part of the index implementation, and I was therefore hopeful that pre-sorting could result in a better distribution of the data across the index tree.

The difficult part in applying pre-sorting to the indexed bit strings is that the implemented linear ordering should preserve the locality in the original search space (items that are close to each other when ordered left-to-right across the leaf nodes, should be also close to each other according to the search criteria; for geometric and low-dimensionality indices this topic has a direct and very interesting connection to that of space-filling curves).

Gray codes are binary encodings where consecutive bitstrings by definition differ in the value for one single bit, and may be therefore assumed to be somewhat similar to each other. Because implementing a comparator function for the "binary reflected gray code" looked relatively easy, I decided to explore this option for the bfp and mol indices.

Apparently, this worked well for substructure searches, in my first tests the index construction is only marginally faster, but queries look faster (40 to 5 % depending on the selectivity - "unfortunately" it's fast queries becoming even faster, but there's a small benefit for low selectivity queries too).

For the bfp index I didn't run any exhaustive tests, but the index construction is for some reason 2x slower and the few search queries I tested didn't perform very differently with pre-sorting enabled. I suspect this may be due to how the binary reflected gray code orders the strings in a way that tends to conflict with criteria that would be instead very useful to similarity queries. Monotonic (or almost monotonic) gray codes exist that sort the bitstrings in increasing weight/popcount order. For this reason, I think a monotonic gray code would probably work better for similarity queries, but implementing a suitable comparator function looks much more challenging (and I also have a couple other topics I'd like to explore re similarity searches).

### CODE REVIEW

**Comment by rvianello** [comment]
In summary, I think at present the main outcome from this exercise is that the `gmol_penalty` and `gmol_picksplit` function don't do a really great job at distributing the entries in the search tree in a way that is ideal to support the substructure search queries, and ordering the data according to a binary reflected gray code might work better. Similarly to the bfp/similarity search case, it's possible that other variants (e.g. a long-run gray code) could work better, but implementing a corresponding comparator looks challenging).

* [Gray code at Wikipedia](https://en.wikipedia.org/wiki/Gray_code)
* [Alternative Gray codes](https://sciyoshi.com/2010/12/gray-codes/)

**Comment by rvianello** [comment]
sorry, are there any known issues affecting the Ubuntu_x64_cartridge build? it looks like it failed at installing the cartridge files with "Permission denied" errors, and consequently failed all tests, but I was comparing this behavior with a different recent PR, and in that case it similarly failed at installing the cartridge and the tests still passed, which is even more surprising to me.

**Comment by ptosco** [comment]
@rvianello Hi Riccardo, I have just checked the logs.
If you go through the `linux_build_catrdige.yml` file you will see that the first failure (due to insufficient privileges at the `make install`) stage is indeed expected since the normal user cannot write to `/usr`:
```
2023-04-23T17:36:25.5941407Z =====================================================================
2023-04-23T17:36:25.5941766Z PostgreSQL cartridge installation failed:
2023-04-23T17:36:25.5942101Z =====================================================================
2023-04-23T17:36:25.5942499Z + test -n ''
2023-04-23T17:36:25.5942951Z + cp /home/vsts/work/1/s/build/Code/PgSQL/rdkit/rdkit--4.3.1.sql /usr/share/postgresql/12/extension
2023-04-23T17:36:25.5943548Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--4.3.1.sql': Permission denied
2023-04-23T17:36:25.5943967Z + cp /home/vsts/work/1/s/Code/PgSQL/rdkit/rdkit.control /usr/share/postgresql/12/extension
2023-04-23T17:36:25.5944442Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit.control': Permission denied
2023-04-23T17:36:25.5945568Z + cp /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--3.8--4.0.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.0--4.0.1.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.0.1--4.1.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.1--4.1.1.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.1.1--4.2.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.2.0--4.3.0.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.3.0--4.3.1.sql /usr/share/postgresql/12/extension
2023-04-23T17:36:25.5947800Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--3.8--4.0.sql': Permission denied
2023-04-23T17:36:25.5948382Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--4.0--4.0.1.sql': Permission denied
2023-04-23T17:36:25.5948895Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--4.0.1--4.1.sql': Permission denied
2023-04-23T17:36:25.5949397Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--4.1--4.1.1.sql': Permission denied
2023-04-23T17:36:25.5949909Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--4.1.1--4.2.sql': Permission denied
2023-04-23T17:36:25.5950422Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--4.2.0--4.3.0.sql': Permission denied
2023-04-23T17:36:25.5950928Z cp: cannot create regular file '/usr/share/postgresql/12/extension/rdkit--4.3.0--4.3.1.sql': Permission denied
2023-04-23T17:36:25.5951324Z + cp /home/vsts/work/1/s/build/Code/PgSQL/rdkit/librdkit.so /usr/lib/postgresql/12/lib/rdkit.so
2023-04-23T17:36:25.5951815Z cp: cannot create regular file '/usr/lib/postgresql/12/lib/rdkit.so': Permission denied
2023-04-23T17:36:25.5952079Z 
2023-04-23T17:36:25.5952392Z =====================================================================
2023-04-23T17:36:25.5952881Z This might be due to insufficient privileges.
2023-04-23T17:36:25.5953243Z Check /home/vsts/work/1/s/build/Code/PgSQL/rdkit/pgsql_install.sh
2023-04-23T17:36:25.5953573Z for correctness of installation paths. If everything is OK, gain 
2023-04-23T17:36:25.5953918Z administrator privileges, stop the PostgreSQL service, run
2023-04-23T17:36:25.5954258Z /home/vsts/work/1/s/build/Code/PgSQL/rdkit/pgsql_install.sh
2023-04-23T17:36:25.5954583Z to install the PostgreSQL RDKit cartridge, then start again
2023-04-23T17:36:25.5954895Z the PostgreSQL service.
2023-04-23T17:36:25.5955206Z =====================================================================
```
However, the subsequent `sudo sh ./Code/PgSQL/rdkit/pgsql_install.sh` succeeds:
```
2023-04-23T17:36:25.6015935Z + test -n 
2023-04-23T17:36:25.6019745Z + cp /home/vsts/work/1/s/build/Code/PgSQL/rdkit/rdkit--4.3.1.sql /usr/share/postgresql/12/extension
2023-04-23T17:36:25.6035967Z + cp /home/vsts/work/1/s/Code/PgSQL/rdkit/rdkit.control /usr/share/postgresql/12/extension
2023-04-23T17:36:25.6054465Z + cp /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--3.8--4.0.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.0--4.0.1.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.0.1--4.1.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.1--4.1.1.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.1.1--4.2.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.2.0--4.3.0.sql /home/vsts/work/1/s/build/Code/PgSQL/rdkit/update_sql/rdkit--4.3.0--4.3.1.sql /usr/share/postgresql/12/extension
2023-04-23T17:36:25.6072604Z + cp /home/vsts/work/1/s/build/Code/PgSQL/rdkit/librdkit.so /usr/lib/postgresql/12/lib/rdkit.so
2023-04-23T17:36:25.6230272Z ##[section]Finishing: Build
```
So `/usr/lib/postgresql/12/lib/rdkit.so` is correctly installed.
Then the tests start:
```
2023-04-23T17:36:26.1606921Z ##[section]Starting: Run tests
```
and fail, which is probably due to something wrong in this cartridge build.
In the recent PR that you checked you have probably been misled by the first installation failure (which is indeed expected).
The fact that all tests passed means that the built cartridge was fully functional.
I hope this makes sense to you.

**Comment by rvianello** [comment]
I suspect something went wrong either in the cmake configuration step, or in the setup of the build environment. I'm not familiar with Ubuntu, but if I understand correctly, in the setup step postgres 14 is uninstalled and replaced by postgres 12. On the other hand, cmake reports `-- Found PostgreSQL: /usr/lib/x86_64-linux-gnu/libpq.so (found version "15.2")`.

At line 50 of the tests ouput I find one line reading: `+ERROR:  invalid function number 11, must be between 1 and 9` that seems to confirm that the code was built against postgres 12, but with the sortsupport method enabled. The inconsistency was probably caused by cmake finding postgres 15.2.

**Comment by ptosco** [comment]
@rvianello Indeed, it looks like
```
    sudo apt-get remove postgresql-14 postgresql-client-14
    sudo apt-get remove postgresql postgresql-common postgresql-client-common
    sudo apt purge postgresql-14 postgresql-client-14
    sudo apt purge postgresql postgresql-common postgresql-client-common
```
did not remove PostgreSQL 15.2 from the system.
My guess is that the above worked when Ubuntu shipped with PostgreSQL 14.
Could you try replacing `-14` with `-15` in `.azure-pipelines/linux_build_cartridge.yml` in your PR? That might fix it.

**Comment by rvianello** [comment]
@ptosco Ah, I see, thanks for the suggestion. I'll try this later today.

**Comment by rvianello** [comment]
@ptosco I briefly tested the changes and then reverted them. It looks like PostgreSQL 15 is not initially present and version 14 is actually removed. Maybe some other package should be pinned when version 12 is installed?

**Comment by ptosco** [comment]
@rvianello I see. Incidentally it looks like the attempt to remove `postgresql` is ineffective:
```
Package 'postgresql' is not installed, so not removed
```
Assuming that PostgreSQL 15 is pulled in as dependency of something else, I would then suggest to explicitly instruct CMake to target PostgreSQL 12 by adding relevant CMake switches:
```
        -D PostgreSQL_INCLUDE_DIR=/usr/include/postgresql/12 \
        -D PostgreSQL_TYPE_INCLUDE_DIR=/usr/include/postgresql/12/server \
        -D PostgreSQL_LIBRARY=/usr/lib/postgresql/12/lib/libpq.so \
```

**Comment by rvianello** [comment]
@ptosco adding libpq5 to the list of removed/purged packages eventually worked.

I'll push another commit (skipping the downgrade to postgresql 12) to actually exercise the modified code.

**Comment by greglandrum** [comment]
@rvianello is this ready to be reviewed?

**Comment by rvianello** [comment]
@greglandrum yes, thank you, I think it is now ready to be reviewed. I repeated my tests with a larger database, which confirmed my earlier results. There should be a performance improvement for substructure search queries, but I by default disabled the pre-sorting support for the bfp gist index, where the implemented sorting criteria is not effective.

I mainly added these cmake options to help with testing, I can eventually remove them again (and cleanup some of the related code) if you prefer to keep the overall build-time configuration simple. 

Also, the CI so far only tested this branch with postgres 12, which is a useful sanity check, but it didn't actually exercise the new code. I don't know if it may be possible to install the missing dev packages for postgres 14, or if there was a different reason for  downgrading it to version 12.

File: Code/PgSQL/rdkit/rdkit_gist.c
Comment (rvianello) [inline Code/PgSQL/rdkit/rdkit_gist.c:560]: the extra copies created by detoasting are usually fine to ignore and assume to be automatically reclaimed, but this comparison function may be called a huge number of times when an index is built, making the explicit deallocation necessary.

File: Code/PgSQL/rdkit/rdkit.h
Comment (greglandrum) [inline Code/PgSQL/rdkit/rdkit.h:45]: Stupid question: why is this in a `do { } while()` that's only going to run once?

**Review by greglandrum — APPROVED** [review]
One minor question here.

To be clear: I haven't actually reviewed that you're using the new postgresql capabilities correctly. I don't have the expertise there and developing it would require a significant chunk of time, so I'm assuming you got that part right. :-)

**Comment by greglandrum** [comment]
> Also, the CI so far only tested this branch with postgres 12, which is a useful sanity check, but it didn't actually exercise the new code. I don't know if it may be possible to install the missing dev packages for postgres 14, or if there was a different reason for downgrading it to version 12.

It's definitely worth doing a bit of exploration to see if we can get the CI to run v14. How about we wait until after this is merged and do that in a separate PR?

File: Code/PgSQL/rdkit/rdkit.h
Comment (rvianello) [reply Code/PgSQL/rdkit/rdkit.h:45]: The do/while loop is just a syntactic trick that allows using the macro more easily in the code, without requiring a different handling of semicolons or curly braces depending on where it's placed. This blog post provides a better explanation and a few examples https://hownot2code.wordpress.com/2016/12/05/do-while-0-in-macros/ (I would have said this construct was already in use somewhere else in RDKit too, but I'm not finding it, I was probably mistaken).

**Comment by rvianello** [comment]
Sounds good to me, thanks.

**Pull request merged**
